<a href="https://colab.research.google.com/github/19003934-UHI/Data_Analytics/blob/main/Assignment_2_TensorFlow_models.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Import and Cleaning




In [35]:
# For data frames
import pandas as pd
# For fast calculations
import numpy as np
# Create data frame from dataset
df = pd.read_csv("https://raw.githubusercontent.com/19003934-UHI/Data_Analytics/main/bquxjob_49d5e1e3_1872d26a08d.csv", index_col=0,)

# Print out data
df[:10]

year  mo  da collision_date  temp  dewp  visib  wdsp  mxpsd   gust   max  \
day                                                                             
7    2018   9  23     2018-09-23  59.8  50.2   10.0   3.0    5.1  999.9  78.1   
4    2018  12  20     2018-12-20  38.6  34.4    9.6   5.0    7.0  999.9  48.0   
2    2013  11   5     2013-11-05  43.5  30.4   10.0   3.9    7.0  999.9  50.0   
3    2012   7  11     2012-07-11  77.1  62.0   10.0   1.9    7.0  999.9  84.0   
4    2012   8   9     2012-08-09  78.2  69.6    9.3   2.3    7.0   15.0  88.0   
4    2012   7  12     2012-07-12  77.4  60.2   10.0   2.3    7.0  999.9  88.0   
2    2017   9  26     2017-09-26  65.9  65.3    2.6   3.4    7.0  999.9  77.0   
1    2012   9  17     2012-09-17  66.5  48.0   10.0   2.2    7.0  999.9  77.0   
4    2020   9  10     2020-09-10  70.8  69.5    0.8   4.4    7.0  999.9  77.0   
1    2014  11  10     2014-11-10  48.2  42.2    9.6   4.0    7.0  999.9  59.0   

      min  prcp   sndp  fog  NUM_COLLISIONS  
day                                          
7    53.1  0.00  999.9    0             475  
4    21.0  0.00  999.9    0             806  
2    37.9  0.00  999.9    0             510  
3    64.9  0.00  999.9    0             565  
4    61.0  0.00  999.9    0             581  
4    57.0  0.00  999.9    0             577  
2    59.0  0.12  999.9    1             656  
1    48.2  0.00  999.9    0             452  
4    66.9  0.00  999.9    1             347  
1    39.0  0.00  999.9    0             560

In [36]:
# Converting data to include indexing for easier manipulation, done for each year between 2012 and 2020.
print(2012)
collisions2012temp = df[df['year'] == 2012]
dayofyear = range(1, len(collisions2012temp)+1)
dayindex = range(1, len(collisions2012temp)+1)
collisions2012 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2012temp.reset_index(drop=True)], axis=1)
collisions2012temp[:10]

2012


year  mo  da collision_date  temp  dewp  visib  wdsp  mxpsd   gust   max  \
day                                                                             
3    2012   7  11     2012-07-11  77.1  62.0   10.0   1.9    7.0  999.9  84.0   
4    2012   8   9     2012-08-09  78.2  69.6    9.3   2.3    7.0   15.0  88.0   
4    2012   7  12     2012-07-12  77.4  60.2   10.0   2.3    7.0  999.9  88.0   
1    2012   9  17     2012-09-17  66.5  48.0   10.0   2.2    7.0  999.9  77.0   
5    2012   8  24     2012-08-24  77.3  60.9    9.3   2.0    7.0  999.9  87.1   
5    2012   7  13     2012-07-13  78.9  60.7    9.9   3.0    7.0  999.9  87.1   
5    2012  10  26     2012-10-26  60.8  53.4   10.0   1.5    7.0  999.9  64.9   
6    2012   7  28     2012-07-28  75.5  69.6    7.0   3.9    8.0   15.9  84.0   
7    2012   7  22     2012-07-22  72.5  59.4   10.0   2.0    8.0  999.9  82.0   
5    2012   9   7     2012-09-07  77.0  67.8    8.7   2.7    8.0  999.9  86.0   

      min  prcp   sndp  fog  NUM_COLLISIONS  
day                                          
3    64.9   0.0  999.9    0             565  
4    61.0   0.0  999.9    0             581  
4    57.0   0.0  999.9    0             577  
1    48.2   0.0  999.9    0             452  
5    66.0   0.0  999.9    0             601  
5    63.0   0.0  999.9    0             603  
5    39.2   0.0  999.9    0             598  
6    68.0   0.0  999.9    1             513  
7    54.0   0.0  999.9    0             498  
5    64.4   0.0  999.9    0             680

In [37]:
print(2013)
collisions2013temp = df[df['year'] == 2013]
dayofyear = range(1, len(collisions2013temp)+1)
dayindex = range(1, len(collisions2013temp)+1)
collisions2013 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2013temp.reset_index(drop=True)], axis=1)
collisions2013['NUM_COLLISIONS'].describe()

2013


count    365.000000
mean     558.175342
std       83.273922
min      257.000000
25%      505.000000
50%      558.000000
75%      613.000000
max      867.000000
Name: NUM_COLLISIONS, dtype: float64

In [38]:
print(2014)
collisions2014temp = df[df['year'] == 2014]
dayofyear = range(1, len(collisions2014temp)+1)
dayindex = range(1, len(collisions2014temp)+1)
collisions2014 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2014temp.reset_index(drop=True)], axis=1)
collisions2014['NUM_COLLISIONS'].describe()

2014


count     424.000000
mean      564.257075
std        90.034285
min       303.000000
25%       512.750000
50%       567.500000
75%       620.250000
max      1161.000000
Name: NUM_COLLISIONS, dtype: float64

In [39]:
print(2015)
collisions2015temp = df[df['year'] == 2015]
dayofyear = range(1, len(collisions2015temp)+1)
dayindex = range(1, len(collisions2015temp)+1)
collisions2015 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2015temp.reset_index(drop=True)], axis=1)
collisions2015['NUM_COLLISIONS'].describe()

2015


count    365.000000
mean     596.421918
std      102.801594
min      188.000000
25%      532.000000
50%      601.000000
75%      666.000000
max      961.000000
Name: NUM_COLLISIONS, dtype: float64

In [40]:
print(2016)
collisions2016temp = df[df['year'] == 2016]
dayofyear = range(1, len(collisions2016temp)+1)
dayindex = range(1, len(collisions2016temp)+1)
collisions2016 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2016temp.reset_index(drop=True)], axis=1)
collisions2016['NUM_COLLISIONS'].describe()

2016


count    366.000000
mean     627.953552
std       97.433957
min      291.000000
25%      568.250000
50%      635.000000
75%      696.750000
max      872.000000
Name: NUM_COLLISIONS, dtype: float64

In [41]:
print(2017)
collisions2017temp = df[df['year'] == 2017]
dayofyear = range(1, len(collisions2017temp)+1)
dayindex = range(1, len(collisions2017temp)+1)
collisions2017 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2017temp.reset_index(drop=True)], axis=1)
collisions2017['NUM_COLLISIONS'].describe()

2017


count    365.000000
mean     632.895890
std      102.339261
min      333.000000
25%      570.000000
50%      636.000000
75%      700.000000
max      999.000000
Name: NUM_COLLISIONS, dtype: float64

In [42]:
print(2018)
collisions2018temp = df[df['year'] == 2018]
dayofyear = range(1, len(collisions2018temp)+1)
dayindex = range(1, len(collisions2018temp)+1)
collisions2018 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2018temp.reset_index(drop=True)], axis=1)
collisions2018['NUM_COLLISIONS'].describe()

2018


count     365.000000
mean      634.421918
std        99.876531
min       328.000000
25%       562.000000
50%       634.000000
75%       706.000000
max      1065.000000
Name: NUM_COLLISIONS, dtype: float64

In [43]:
print(2019)
collisions2019temp = df[df['year'] == 2019]
dayofyear = range(1, len(collisions2019temp)+1)
dayindex = range(1, len(collisions2019temp)+1)
collisions2019 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2019temp.reset_index(drop=True)], axis=1)
collisions2019['NUM_COLLISIONS'].describe()

2019


count    365.000000
mean     579.413699
std       88.947085
min      264.000000
25%      522.000000
50%      589.000000
75%      641.000000
max      803.000000
Name: NUM_COLLISIONS, dtype: float64

In [44]:
print(2020)
collisions2020temp = df[df['year'] == 2020]
dayofyear = range(1, len(collisions2020temp)+1)
dayindex = range(1, len(collisions2020temp)+1)
collisions2020 = pd.concat([pd.DataFrame(dayindex), pd.DataFrame(dayofyear), collisions2020temp.reset_index(drop=True)], axis=1)
collisions2020['NUM_COLLISIONS'].describe()

2020


count    366.000000
mean     308.510929
std      110.188996
min       94.000000
25%      242.250000
50%      301.000000
75%      348.000000
max      774.000000
Name: NUM_COLLISIONS, dtype: float64

In [45]:
# Add years together into one dataset
collisions_2012_to_2020 = pd.concat([collisions2012, collisions2013, collisions2014, collisions2015, collisions2016, collisions2017, collisions2018, collisions2019, collisions2020], axis=0)
print(collisions_2012_to_2020.head())

df = collisions_2012_to_2020

   0  0  year  mo  da collision_date  temp  dewp  visib  wdsp  mxpsd   gust  \
0  1  1  2012   7  11     2012-07-11  77.1  62.0   10.0   1.9    7.0  999.9   
1  2  2  2012   8   9     2012-08-09  78.2  69.6    9.3   2.3    7.0   15.0   
2  3  3  2012   7  12     2012-07-12  77.4  60.2   10.0   2.3    7.0  999.9   
3  4  4  2012   9  17     2012-09-17  66.5  48.0   10.0   2.2    7.0  999.9   
4  5  5  2012   8  24     2012-08-24  77.3  60.9    9.3   2.0    7.0  999.9   

    max   min  prcp   sndp  fog  NUM_COLLISIONS  
0  84.0  64.9   0.0  999.9    0             565  
1  88.0  61.0   0.0  999.9    0             581  
2  88.0  57.0   0.0  999.9    0             577  
3  77.0  48.2   0.0  999.9    0             452  
4  87.1  66.0   0.0  999.9    0             601  


In [46]:
# Imports for creating validation sets and linear regression model
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt

#Tensorflow Setup

In [47]:
# Set df equal to the entire collision set
df = collisions_2012_to_2020
print(df[:10])

# needed to create the data frame
import pandas as pd

# needed to help with speedy maths based calculations
import numpy as np

import tensorflow as tf

from tensorflow import keras
from tensorflow.keras import layers

     

    0   0  year  mo  da collision_date  temp  dewp  visib  wdsp  mxpsd   gust  \
0   1   1  2012   7  11     2012-07-11  77.1  62.0   10.0   1.9    7.0  999.9   
1   2   2  2012   8   9     2012-08-09  78.2  69.6    9.3   2.3    7.0   15.0   
2   3   3  2012   7  12     2012-07-12  77.4  60.2   10.0   2.3    7.0  999.9   
3   4   4  2012   9  17     2012-09-17  66.5  48.0   10.0   2.2    7.0  999.9   
4   5   5  2012   8  24     2012-08-24  77.3  60.9    9.3   2.0    7.0  999.9   
5   6   6  2012   7  13     2012-07-13  78.9  60.7    9.9   3.0    7.0  999.9   
6   7   7  2012  10  26     2012-10-26  60.8  53.4   10.0   1.5    7.0  999.9   
7   8   8  2012   7  28     2012-07-28  75.5  69.6    7.0   3.9    8.0   15.9   
8   9   9  2012   7  22     2012-07-22  72.5  59.4   10.0   2.0    8.0  999.9   
9  10  10  2012   9   7     2012-09-07  77.0  67.8    8.7   2.7    8.0  999.9   

    max   min  prcp   sndp  fog  NUM_COLLISIONS  
0  84.0  64.9   0.0  999.9    0             565  
1  88.0 

The first ten values of the dataset will be used in training the TensorFlow models.

##TensorFlow Version 1

Initial Settings:
*   df_dnn_input - axis = 1
*   training_dataset - frac = 0.8, random_state = 0
*   SCALE_NUM_COLLISIONS = 0.8
*   normaliser - axis = -1
*   dnn_model_1 - normaliser, 2 layers of 48, using relu, and a layer of 1
*   dnn_model_1 compile - Adam optimizer, of 0.001
*   history - validation_split = 0.8, epochs = 100


In [48]:
dnn_input_data = [df["year"], df["mo"], df["da"], df["temp"], df["dewp"], df["visib"], df["wdsp"], df["mxpsd"], df["gust"], df["max"], df["min"], df["prcp"], df["sndp"], df["fog"], df["NUM_COLLISIONS"]]
headers = ["year","mo","da","temp","dewp","visib","wdsp","mxpsd","gust","max","min","prcp","sndp","fog","NUM_COLLISIONS"]
df_dnn_input = pd.concat(dnn_input_data, axis=1, keys=headers)

training_dataset = df_dnn_input.sample(frac=0.8, random_state=0)
test_dataset = df_dnn_input.drop(training_dataset.index)
     
training_features = training_dataset.copy()
test_features = test_dataset.copy()

training_labels = training_features.pop('NUM_COLLISIONS')
test_labels = test_features.pop('NUM_COLLISIONS')
     
SCALE_NUM_COLLISIONS = 0.8
     
training_labels = training_labels/SCALE_NUM_COLLISIONS
test_labels = test_labels/SCALE_NUM_COLLISIONS

normaliser = tf.keras.layers.Normalization(axis=-1)
normaliser.adapt(np.array(training_features))

# This is the only difference, instead of a single layer, we have our normalisation layer (22 inputs), 2 layers of 48, with 1 output. The 48 can be adjusted to improve the net.
dnn_model_1 = keras.Sequential([
      normaliser,
      layers.Dense(48, activation='relu'),
      layers.Dense(48, activation='relu'),
      layers.Dense(1)
  ])

dnn_model_1.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.001))

In [49]:
%%time
history = dnn_model_1.fit(
    training_features,
    training_labels,
    validation_split=0.8,
    verbose=0, 
    epochs=100)

# remember, we want to minimise this. The model with the lowest is the best. 
dnn_model_1_results = dnn_model_1.evaluate(test_features, test_labels, verbose=0)
print(dnn_model_1_results)

101.15771484375
CPU times: user 15.3 s, sys: 992 ms, total: 16.3 s
Wall time: 21.3 s


In [50]:
# make sure the labels match up with the dataframe from earlier.
input_1 = pd.DataFrame.from_dict(data = 
				{
          'year' : [2012,2012,2012,2012,2012,2012,2012,2012,2012,2012],
          'mo' : [7,8,7,9,8,7,10,7,7,9],
          'da' : [11,9,12,17,24,13,26,28,22,7],
          'temp' : [77.1,78.2,77.4,66.5,77.3,78.9,60.8,75.5,72.5,77.0],
          'dewp' : [62.0,69.6,60.2,48.0,60.9,60.7,53.4,69.6,59.4,67.8],
          'visib' : [10.0,9.9,10.0,10.0,9.3,9.9,10.0,7.0,10.0,8.7],
          'wdsp' : [1.9,2.3,2.3,2.2,2.0,3.0,1.5,3.9,2.0,2.7],
          'mxpsd' : [7.0,7.0,7.0,7.0,7.0,7.0,7.0,8.0,8.0,8.0],
          'gust' : [999.9,15.0,999.9,999.9,999.9,999.9,999.9,15.9,999.9,999.9],
          'max' : [84.0,88.0,88.0,77.0,87.1,87.1,64.9,84.0,82.0,86.0],
          'min' : [64.9,61.0,57.0,48.2,66.0,63.0,39.2,68.0,54.0,64.4],
          'prcp' : [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
          'sndp' : [999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9],
          'fog' : [0,0,0,0,0,0,0,1,0,0],
        })
linear_collision_predictions_1 = dnn_model_1.predict(input_1[:10])*SCALE_NUM_COLLISIONS
linear_collision_predictions_1

1/1 [==============================] - 0s 66ms/step


array([[651.5684 ],
       [626.47577],
       [659.45935],
       [644.8503 ],
       [672.8525 ],
       [659.2323 ],
       [644.07367],
       [664.6025 ],
       [654.9274 ],
       [633.8153 ]], dtype=float32)

Results: 
*   dnn_model_1_results = 101.15771484375
*   linear_collision_predictions: 
        [[651.5684 ],
        [626.47577],
        [659.45935],
        [644.8503 ],
        [672.8525 ],
        [659.2323 ],
        [644.07367],
        [664.6025 ],
        [654.9274 ],
        [633.8153 ]]


As seen below, the resulting collision rates the model predicts are higher than the actual data. The next model will aim to lower the rate of collision.


In [51]:
df_dnn_input[:10]

year  mo  da  temp  dewp  visib  wdsp  mxpsd   gust   max   min  prcp  \
0  2012   7  11  77.1  62.0   10.0   1.9    7.0  999.9  84.0  64.9   0.0   
1  2012   8   9  78.2  69.6    9.3   2.3    7.0   15.0  88.0  61.0   0.0   
2  2012   7  12  77.4  60.2   10.0   2.3    7.0  999.9  88.0  57.0   0.0   
3  2012   9  17  66.5  48.0   10.0   2.2    7.0  999.9  77.0  48.2   0.0   
4  2012   8  24  77.3  60.9    9.3   2.0    7.0  999.9  87.1  66.0   0.0   
5  2012   7  13  78.9  60.7    9.9   3.0    7.0  999.9  87.1  63.0   0.0   
6  2012  10  26  60.8  53.4   10.0   1.5    7.0  999.9  64.9  39.2   0.0   
7  2012   7  28  75.5  69.6    7.0   3.9    8.0   15.9  84.0  68.0   0.0   
8  2012   7  22  72.5  59.4   10.0   2.0    8.0  999.9  82.0  54.0   0.0   
9  2012   9   7  77.0  67.8    8.7   2.7    8.0  999.9  86.0  64.4   0.0   

    sndp  fog  NUM_COLLISIONS  
0  999.9    0             565  
1  999.9    0             581  
2  999.9    0             577  
3  999.9    0             452  
4  999.9    0             601  
5  999.9    0             603  
6  999.9    0             598  
7  999.9    1             513  
8  999.9    0             498  
9  999.9    0             680

##TensorFlow Version 2

Changed Settings:
*   training_dataset - frac = 0.5, random_state = 64
*   SCALE_NUM_COLLISIONS = 0.75


In [52]:
dnn_input_data = [df["year"], df["mo"], df["da"], df["temp"], df["dewp"], df["visib"], df["wdsp"], df["mxpsd"], df["gust"], df["max"], df["min"], df["prcp"], df["sndp"], df["fog"], df["NUM_COLLISIONS"]]
headers = ["year","mo","da","temp","dewp","visib","wdsp","mxpsd","gust","max","min","prcp","sndp","fog","NUM_COLLISIONS"]
df_dnn_input = pd.concat(dnn_input_data, axis=1, keys=headers)

training_dataset = df_dnn_input.sample(frac=0.5, random_state=64)
test_dataset = df_dnn_input.drop(training_dataset.index)
     
training_features = training_dataset.copy()
test_features = test_dataset.copy()

training_labels = training_features.pop('NUM_COLLISIONS')
test_labels = test_features.pop('NUM_COLLISIONS')
     

# A scale is not required here, but the constant will be useful in the assignment.
SCALE_NUM_COLLISIONS = 0.75
     
training_labels = training_labels/SCALE_NUM_COLLISIONS
test_labels = test_labels/SCALE_NUM_COLLISIONS

normaliser = tf.keras.layers.Normalization(axis=-1)
normaliser.adapt(np.array(training_features))

# This is the only difference, instead of a single layer, we have our normalisation layer (22 inputs), 2 layers of 48, with 1 output. The 48 can be adjusted to improve the net.
dnn_model_2 = keras.Sequential([
      normaliser,
      layers.Dense(48, activation='relu'),
      layers.Dense(48, activation='relu'),
      layers.Dense(1)
  ])

dnn_model_2.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.001))

In [53]:
%%time
history = dnn_model_2.fit(
    training_features,
    training_labels,
    validation_split=0.8,
    verbose=0, 
    epochs=100)

# remember, we want to minimise this. The model with the lowest is the best. 
dnn_model_2_results = dnn_model_2.evaluate(test_features, test_labels, verbose=0)
print(dnn_model_2_results)

389.8008117675781
CPU times: user 10.8 s, sys: 657 ms, total: 11.4 s
Wall time: 11.4 s


In [54]:
# make sure the labels match up with the dataframe from earlier.
input_2 = pd.DataFrame.from_dict(data = 
				{
          'year' : [2012,2012,2012,2012,2012,2012,2012,2012,2012,2012],
          'mo' : [7,8,7,9,8,7,10,7,7,9],
          'da' : [11,9,12,17,24,13,26,28,22,7],
          'temp' : [77.1,78.2,77.4,66.5,77.3,78.9,60.8,75.5,72.5,77.0],
          'dewp' : [62.0,69.6,60.2,48.0,60.9,60.7,53.4,69.6,59.4,67.8],
          'visib' : [10.0,9.9,10.0,10.0,9.3,9.9,10.0,7.0,10.0,8.7],
          'wdsp' : [1.9,2.3,2.3,2.2,2.0,3.0,1.5,3.9,2.0,2.7],
          'mxpsd' : [7.0,7.0,7.0,7.0,7.0,7.0,7.0,8.0,8.0,8.0],
          'gust' : [999.9,15.0,999.9,999.9,999.9,999.9,999.9,15.9,999.9,999.9],
          'max' : [84.0,88.0,88.0,77.0,87.1,87.1,64.9,84.0,82.0,86.0],
          'min' : [64.9,61.0,57.0,48.2,66.0,63.0,39.2,68.0,54.0,64.4],
          'prcp' : [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
          'sndp' : [999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9],
          'fog' : [0,0,0,0,0,0,0,1,0,0],
        })
linear_collision_predictions_2 = dnn_model_2.predict(input_2[:10])*SCALE_NUM_COLLISIONS
linear_collision_predictions_2

1/1 [==============================] - 0s 75ms/step


array([[693.7556 ],
       [680.27936],
       [695.4418 ],
       [672.5981 ],
       [707.3334 ],
       [692.5575 ],
       [674.19885],
       [655.9875 ],
       [669.8555 ],
       [688.11865]], dtype=float32)

Results: 
*   dnn_model_2_results = 226.68710327148438
*   linear_collision_predictions: 
        [[693.7556 ],
        [680.27936],
        [695.4418 ],
        [672.5981 ],
        [707.3334 ],
        [692.5575 ],
        [674.19885],
        [655.9875 ],
        [669.8555 ],
        [688.11865]]


As seen below, the resulting collision rates the model predicts are still higher than the actual data even when changing some settings. The next model will further aim to lower the rate of collision.


In [55]:
df_dnn_input[:10]

year  mo  da  temp  dewp  visib  wdsp  mxpsd   gust   max   min  prcp  \
0  2012   7  11  77.1  62.0   10.0   1.9    7.0  999.9  84.0  64.9   0.0   
1  2012   8   9  78.2  69.6    9.3   2.3    7.0   15.0  88.0  61.0   0.0   
2  2012   7  12  77.4  60.2   10.0   2.3    7.0  999.9  88.0  57.0   0.0   
3  2012   9  17  66.5  48.0   10.0   2.2    7.0  999.9  77.0  48.2   0.0   
4  2012   8  24  77.3  60.9    9.3   2.0    7.0  999.9  87.1  66.0   0.0   
5  2012   7  13  78.9  60.7    9.9   3.0    7.0  999.9  87.1  63.0   0.0   
6  2012  10  26  60.8  53.4   10.0   1.5    7.0  999.9  64.9  39.2   0.0   
7  2012   7  28  75.5  69.6    7.0   3.9    8.0   15.9  84.0  68.0   0.0   
8  2012   7  22  72.5  59.4   10.0   2.0    8.0  999.9  82.0  54.0   0.0   
9  2012   9   7  77.0  67.8    8.7   2.7    8.0  999.9  86.0  64.4   0.0   

    sndp  fog  NUM_COLLISIONS  
0  999.9    0             565  
1  999.9    0             581  
2  999.9    0             577  
3  999.9    0             452  
4  999.9    0             601  
5  999.9    0             603  
6  999.9    0             598  
7  999.9    1             513  
8  999.9    0             498  
9  999.9    0             680

##TensorFlow Version 3

Changed Settings:
*   training_dataset - frac = 0.8, random_state = 32
*   SCALE_NUM_COLLISIONS = 1.0
*   dnn_model_3 - layer of 64 instead of two layers of 48
*   history - validation_split = 0.1, epochs = 150


In [56]:
dnn_input_data = [df["year"], df["mo"], df["da"], df["temp"], df["dewp"], df["visib"], df["wdsp"], df["mxpsd"], df["gust"], df["max"], df["min"], df["prcp"], df["sndp"], df["fog"], df["NUM_COLLISIONS"]]
headers = ["year","mo","da","temp","dewp","visib","wdsp","mxpsd","gust","max","min","prcp","sndp","fog","NUM_COLLISIONS"]
df_dnn_input = pd.concat(dnn_input_data, axis=1, keys=headers)

training_dataset = df_dnn_input.sample(frac=0.8, random_state=32)
test_dataset = df_dnn_input.drop(training_dataset.index)
     
training_features = training_dataset.copy()
test_features = test_dataset.copy()

training_labels = training_features.pop('NUM_COLLISIONS')
test_labels = test_features.pop('NUM_COLLISIONS')
     
SCALE_NUM_COLLISIONS = 1.0
     
training_labels = training_labels/SCALE_NUM_COLLISIONS
test_labels = test_labels/SCALE_NUM_COLLISIONS

normaliser = tf.keras.layers.Normalization(axis=-1)
normaliser.adapt(np.array(training_features))

dnn_model_3 = keras.Sequential([
      normaliser,
      layers.Dense(64, activation='relu'),
      layers.Dense(48, activation='relu'),
      layers.Dense(1)
  ])

dnn_model_3.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.001))

In [57]:
%%time
history = dnn_model_3.fit(
    training_features,
    training_labels,
    validation_split=0.1,
    verbose=0, 
    epochs=150)

dnn_model_3_results = dnn_model_3.evaluate(test_features, test_labels, verbose=0)
print(dnn_model_3_results)

73.05980682373047
CPU times: user 25.3 s, sys: 1.35 s, total: 26.6 s
Wall time: 41.8 s


In [58]:
# make sure the labels match up with the dataframe from earlier.
input_3 = pd.DataFrame.from_dict(data = 
				{
          'year' : [2012,2012,2012,2012,2012,2012,2012,2012,2012,2012],
          'mo' : [7,8,7,9,8,7,10,7,7,9],
          'da' : [11,9,12,17,24,13,26,28,22,7],
          'temp' : [77.1,78.2,77.4,66.5,77.3,78.9,60.8,75.5,72.5,77.0],
          'dewp' : [62.0,69.6,60.2,48.0,60.9,60.7,53.4,69.6,59.4,67.8],
          'visib' : [10.0,9.9,10.0,10.0,9.3,9.9,10.0,7.0,10.0,8.7],
          'wdsp' : [1.9,2.3,2.3,2.2,2.0,3.0,1.5,3.9,2.0,2.7],
          'mxpsd' : [7.0,7.0,7.0,7.0,7.0,7.0,7.0,8.0,8.0,8.0],
          'gust' : [999.9,15.0,999.9,999.9,999.9,999.9,999.9,15.9,999.9,999.9],
          'max' : [84.0,88.0,88.0,77.0,87.1,87.1,64.9,84.0,82.0,86.0],
          'min' : [64.9,61.0,57.0,48.2,66.0,63.0,39.2,68.0,54.0,64.4],
          'prcp' : [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
          'sndp' : [999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9],
          'fog' : [0,0,0,0,0,0,0,1,0,0],
        })
linear_collision_predictions_3 = dnn_model_3.predict(input_3[:10])*SCALE_NUM_COLLISIONS
linear_collision_predictions_3

1/1 [==============================] - 0s 65ms/step


array([[584.16437],
       [619.23425],
       [600.4199 ],
       [576.6481 ],
       [582.5792 ],
       [588.7636 ],
       [578.8091 ],
       [606.42285],
       [589.5628 ],
       [611.75854]], dtype=float32)

Results: 
*   dnn_model_3_results = 73.05980682373047
*   linear_collision_predictions: 
        [[584.16437],
        [619.23425],
        [600.4199 ],
        [576.6481 ],
        [582.5792 ],
        [588.7636 ],
        [578.8091 ],
        [606.42285],
        [589.5628 ],
        [611.75854]]

By decreasing the random state and increasing the frac value on the training dataset, it seems to have brought the model more in line with the rate of collisions present in the df_dnn_input model.


In [59]:
df_dnn_input[:10]

year  mo  da  temp  dewp  visib  wdsp  mxpsd   gust   max   min  prcp  \
0  2012   7  11  77.1  62.0   10.0   1.9    7.0  999.9  84.0  64.9   0.0   
1  2012   8   9  78.2  69.6    9.3   2.3    7.0   15.0  88.0  61.0   0.0   
2  2012   7  12  77.4  60.2   10.0   2.3    7.0  999.9  88.0  57.0   0.0   
3  2012   9  17  66.5  48.0   10.0   2.2    7.0  999.9  77.0  48.2   0.0   
4  2012   8  24  77.3  60.9    9.3   2.0    7.0  999.9  87.1  66.0   0.0   
5  2012   7  13  78.9  60.7    9.9   3.0    7.0  999.9  87.1  63.0   0.0   
6  2012  10  26  60.8  53.4   10.0   1.5    7.0  999.9  64.9  39.2   0.0   
7  2012   7  28  75.5  69.6    7.0   3.9    8.0   15.9  84.0  68.0   0.0   
8  2012   7  22  72.5  59.4   10.0   2.0    8.0  999.9  82.0  54.0   0.0   
9  2012   9   7  77.0  67.8    8.7   2.7    8.0  999.9  86.0  64.4   0.0   

    sndp  fog  NUM_COLLISIONS  
0  999.9    0             565  
1  999.9    0             581  
2  999.9    0             577  
3  999.9    0             452  
4  999.9    0             601  
5  999.9    0             603  
6  999.9    0             598  
7  999.9    1             513  
8  999.9    0             498  
9  999.9    0             680

##TensorFlow Version 4

Changed Settings:
*   Increased the number of layers in the network, added one more layer of 128 using relu.


In [60]:
dnn_input_data = [df["year"], df["mo"], df["da"], df["temp"], df["dewp"], df["visib"], df["wdsp"], df["mxpsd"], df["gust"], df["max"], df["min"], df["prcp"], df["sndp"], df["fog"], df["NUM_COLLISIONS"]]
headers = ["year","mo","da","temp","dewp","visib","wdsp","mxpsd","gust","max","min","prcp","sndp","fog","NUM_COLLISIONS"]
df_dnn_input = pd.concat(dnn_input_data, axis=1, keys=headers)

training_dataset = df_dnn_input.sample(frac=0.8, random_state=32)
test_dataset = df_dnn_input.drop(training_dataset.index)
     
training_features = training_dataset.copy()
test_features = test_dataset.copy()

training_labels = training_features.pop('NUM_COLLISIONS')
test_labels = test_features.pop('NUM_COLLISIONS')
     
SCALE_NUM_COLLISIONS = 1.0
     
training_labels = training_labels/SCALE_NUM_COLLISIONS
test_labels = test_labels/SCALE_NUM_COLLISIONS

normaliser = tf.keras.layers.Normalization(axis=-1)
normaliser.adapt(np.array(training_features))

dnn_model_4 = keras.Sequential([
    normaliser,
    layers.Dense(128, activation='relu'),
    layers.Dense(64, activation='relu'),
    layers.Dense(48, activation='relu'),
    layers.Dense(1)
])
dnn_model_4.compile(loss='mean_absolute_error',
                optimizer=tf.keras.optimizers.Adam(0.001))

In [61]:
%%time
history = dnn_model_4.fit(
    training_features,
    training_labels,
    validation_split=0.1,
    verbose=0, 
    epochs=150)

dnn_model_4_results = dnn_model_4.evaluate(test_features, test_labels, verbose=0)
print(dnn_model_4_results)

64.13447570800781
CPU times: user 31.3 s, sys: 1.35 s, total: 32.6 s
Wall time: 31.2 s


In [62]:
# make sure the labels match up with the dataframe from earlier.
input_4 = pd.DataFrame.from_dict(data = 
				{
          'year' : [2012,2012,2012,2012,2012,2012,2012,2012,2012,2012],
          'mo' : [7,8,7,9,8,7,10,7,7,9],
          'da' : [11,9,12,17,24,13,26,28,22,7],
          'temp' : [77.1,78.2,77.4,66.5,77.3,78.9,60.8,75.5,72.5,77.0],
          'dewp' : [62.0,69.6,60.2,48.0,60.9,60.7,53.4,69.6,59.4,67.8],
          'visib' : [10.0,9.9,10.0,10.0,9.3,9.9,10.0,7.0,10.0,8.7],
          'wdsp' : [1.9,2.3,2.3,2.2,2.0,3.0,1.5,3.9,2.0,2.7],
          'mxpsd' : [7.0,7.0,7.0,7.0,7.0,7.0,7.0,8.0,8.0,8.0],
          'gust' : [999.9,15.0,999.9,999.9,999.9,999.9,999.9,15.9,999.9,999.9],
          'max' : [84.0,88.0,88.0,77.0,87.1,87.1,64.9,84.0,82.0,86.0],
          'min' : [64.9,61.0,57.0,48.2,66.0,63.0,39.2,68.0,54.0,64.4],
          'prcp' : [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
          'sndp' : [999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9],
          'fog' : [0,0,0,0,0,0,0,1,0,0],
        })
linear_collision_predictions_4 = dnn_model_4.predict(input_4[:10])*SCALE_NUM_COLLISIONS
linear_collision_predictions_4

1/1 [==============================] - 0s 90ms/step


array([[567.5232 ],
       [578.8622 ],
       [577.18054],
       [540.3776 ],
       [550.1647 ],
       [561.58264],
       [537.2595 ],
       [554.8867 ],
       [546.68823],
       [598.48596]], dtype=float32)

Results: 
*   dnn_model_4_results = 64.13447570800781
*   linear_collision_predictions: 
        [[567.5232 ],
        [578.8622 ],
        [577.18054],
        [540.3776 ],
        [550.1647 ],
        [561.58264],
        [537.2595 ],
        [554.8867 ],
        [546.68823],
        [598.48596]]
Adding another layer to the network helped to bring the prediction closer inline to the collisions of the orginal data, so this settings will be kept while the model gets further refined.


In [63]:
df_dnn_input[:10]

year  mo  da  temp  dewp  visib  wdsp  mxpsd   gust   max   min  prcp  \
0  2012   7  11  77.1  62.0   10.0   1.9    7.0  999.9  84.0  64.9   0.0   
1  2012   8   9  78.2  69.6    9.3   2.3    7.0   15.0  88.0  61.0   0.0   
2  2012   7  12  77.4  60.2   10.0   2.3    7.0  999.9  88.0  57.0   0.0   
3  2012   9  17  66.5  48.0   10.0   2.2    7.0  999.9  77.0  48.2   0.0   
4  2012   8  24  77.3  60.9    9.3   2.0    7.0  999.9  87.1  66.0   0.0   
5  2012   7  13  78.9  60.7    9.9   3.0    7.0  999.9  87.1  63.0   0.0   
6  2012  10  26  60.8  53.4   10.0   1.5    7.0  999.9  64.9  39.2   0.0   
7  2012   7  28  75.5  69.6    7.0   3.9    8.0   15.9  84.0  68.0   0.0   
8  2012   7  22  72.5  59.4   10.0   2.0    8.0  999.9  82.0  54.0   0.0   
9  2012   9   7  77.0  67.8    8.7   2.7    8.0  999.9  86.0  64.4   0.0   

    sndp  fog  NUM_COLLISIONS  
0  999.9    0             565  
1  999.9    0             581  
2  999.9    0             577  
3  999.9    0             452  
4  999.9    0             601  
5  999.9    0             603  
6  999.9    0             598  
7  999.9    1             513  
8  999.9    0             498  
9  999.9    0             680

##TensorFlow Version 5

Changed Settings:
*   Changed Adam learning rate to 0.0001 (was 0.001)
*   Added regularization to the model


In [64]:
from tensorflow.keras import regularizers

dnn_input_data = [df["year"], df["mo"], df["da"], df["temp"], df["dewp"], df["visib"], df["wdsp"], df["mxpsd"], df["gust"], df["max"], df["min"], df["prcp"], df["sndp"], df["fog"], df["NUM_COLLISIONS"]]
headers = ["year","mo","da","temp","dewp","visib","wdsp","mxpsd","gust","max","min","prcp","sndp","fog","NUM_COLLISIONS"]
df_dnn_input = pd.concat(dnn_input_data, axis=1, keys=headers)

training_dataset = df_dnn_input.sample(frac=0.8, random_state=32)
test_dataset = df_dnn_input.drop(training_dataset.index)
     
training_features = training_dataset.copy()
test_features = test_dataset.copy()

training_labels = training_features.pop('NUM_COLLISIONS')
test_labels = test_features.pop('NUM_COLLISIONS')
     
SCALE_NUM_COLLISIONS = 1.0
     
training_labels = training_labels/SCALE_NUM_COLLISIONS
test_labels = test_labels/SCALE_NUM_COLLISIONS

normaliser = tf.keras.layers.Normalization(axis=-1)
normaliser.adapt(np.array(training_features))

dnn_model_5 = keras.Sequential([
    normaliser,
    layers.Dense(128, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(48, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(1)
])
dnn_model_5.compile(loss='mean_absolute_error',
                    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001))

In [65]:
%%time
history = dnn_model_5.fit(
    training_features,
    training_labels,
    validation_split=0.01,
    verbose=0, 
    epochs=150)

dnn_model_5_results = dnn_model_5.evaluate(test_features, test_labels, verbose=0)
print(dnn_model_5_results)

57.25815200805664
CPU times: user 29.8 s, sys: 1.46 s, total: 31.2 s
Wall time: 26.6 s


In [66]:
# make sure the labels match up with the dataframe from earlier.
input_5 = pd.DataFrame.from_dict(data = 
				{
          'year' : [2012,2012,2012,2012,2012,2012,2012,2012,2012,2012],
          'mo' : [7,8,7,9,8,7,10,7,7,9],
          'da' : [11,9,12,17,24,13,26,28,22,7],
          'temp' : [77.1,78.2,77.4,66.5,77.3,78.9,60.8,75.5,72.5,77.0],
          'dewp' : [62.0,69.6,60.2,48.0,60.9,60.7,53.4,69.6,59.4,67.8],
          'visib' : [10.0,9.9,10.0,10.0,9.3,9.9,10.0,7.0,10.0,8.7],
          'wdsp' : [1.9,2.3,2.3,2.2,2.0,3.0,1.5,3.9,2.0,2.7],
          'mxpsd' : [7.0,7.0,7.0,7.0,7.0,7.0,7.0,8.0,8.0,8.0],
          'gust' : [999.9,15.0,999.9,999.9,999.9,999.9,999.9,15.9,999.9,999.9],
          'max' : [84.0,88.0,88.0,77.0,87.1,87.1,64.9,84.0,82.0,86.0],
          'min' : [64.9,61.0,57.0,48.2,66.0,63.0,39.2,68.0,54.0,64.4],
          'prcp' : [0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],
          'sndp' : [999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9,999.9],
          'fog' : [0,0,0,0,0,0,0,1,0,0],
        })
linear_collision_predictions_5 = dnn_model_5.predict(input_5[:10])*SCALE_NUM_COLLISIONS
linear_collision_predictions_5

1/1 [==============================] - 0s 77ms/step


array([[551.6657 ],
       [546.5131 ],
       [569.9788 ],
       [560.2954 ],
       [554.85095],
       [560.2471 ],
       [532.5959 ],
       [546.18427],
       [549.3337 ],
       [546.92267]], dtype=float32)

Results: 
*   dnn_model_5_results = 57.25815200805664
*   linear_collision_predictions: 
        [[551.6657 ],
        [546.5131 ],
        [569.9788 ],
        [560.2954 ],
        [554.85095],
        [560.2471 ],
        [532.5959 ],
        [546.18427],
        [549.3337 ],
        [546.92267]]

By adding regularisation to the model, as well as decreasing the learning rate of the Adam optimiser, it seems to have a higher evaluation score buy lower the rate of collision.


In [67]:
df_dnn_input[:10]

year  mo  da  temp  dewp  visib  wdsp  mxpsd   gust   max   min  prcp  \
0  2012   7  11  77.1  62.0   10.0   1.9    7.0  999.9  84.0  64.9   0.0   
1  2012   8   9  78.2  69.6    9.3   2.3    7.0   15.0  88.0  61.0   0.0   
2  2012   7  12  77.4  60.2   10.0   2.3    7.0  999.9  88.0  57.0   0.0   
3  2012   9  17  66.5  48.0   10.0   2.2    7.0  999.9  77.0  48.2   0.0   
4  2012   8  24  77.3  60.9    9.3   2.0    7.0  999.9  87.1  66.0   0.0   
5  2012   7  13  78.9  60.7    9.9   3.0    7.0  999.9  87.1  63.0   0.0   
6  2012  10  26  60.8  53.4   10.0   1.5    7.0  999.9  64.9  39.2   0.0   
7  2012   7  28  75.5  69.6    7.0   3.9    8.0   15.9  84.0  68.0   0.0   
8  2012   7  22  72.5  59.4   10.0   2.0    8.0  999.9  82.0  54.0   0.0   
9  2012   9   7  77.0  67.8    8.7   2.7    8.0  999.9  86.0  64.4   0.0   

    sndp  fog  NUM_COLLISIONS  
0  999.9    0             565  
1  999.9    0             581  
2  999.9    0             577  
3  999.9    0             452  
4  999.9    0             601  
5  999.9    0             603  
6  999.9    0             598  
7  999.9    1             513  
8  999.9    0             498  
9  999.9    0             680

Of the five models, the changes made to model 4 proved to be the best, and therefore will be used for actual data to see how accurately it predicates the rate of collision.

## Proper TensorFlow Model using Validation Set

In [78]:
# Define the input and output variables
input_vars = [df['year'], df['mo'], df['da'], df['temp'], df['dewp'], df['visib'], df['wdsp'], df['mxpsd'], df['gust'], df['max'], df['min'], df['prcp'], df['sndp'], df['fog']]
output_var = 'NUM_COLLISIONS'

# Create the training and validation datasets
train_data = df.sample(frac=0.8, random_state=32)
val_data = df.drop(train_data.index)

# Split the training and validation datasets into features and labels
train_features = train_data[['max', 'min', 'da']]
train_labels = train_data[output_var]
val_features = val_data[['max', 'min', 'da']]
val_labels = val_data[output_var]

# Scale the output variable
SCALE_NUM_COLLISIONS = 1.0
train_labels = train_labels / SCALE_NUM_COLLISIONS
val_labels = val_labels / SCALE_NUM_COLLISIONS

# Normalize the input features
normalizer = layers.Normalization(axis=-1)
normalizer.adapt(np.array(train_features))

# Define the model architecture
model = keras.Sequential([
    normalizer,
    layers.Dense(128, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(64, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(48, activation='relu', kernel_regularizer=regularizers.l2(0.001)),
    layers.Dense(1)
])

# Compile the model
model.compile(loss='mean_absolute_error',
              optimizer=tf.keras.optimizers.Adam(learning_rate=0.001))

# Train the model
history = model.fit(
    train_features,
    train_labels,
    validation_split=0.1,
    validation_data=(val_features, val_labels),
    epochs=150,
    verbose=1
)

# Evaluate the model
test_data = df.drop(train_data.index)
test_features = test_data[['max', 'min', 'da']]
test_labels = test_data[output_var]
test_labels = test_labels / SCALE_NUM_COLLISIONS
test_loss = model.evaluate(test_features, test_labels, verbose=1)
print(f"Test loss: {test_loss:.4f}")

Epoch 1/150
80/80 [==============================] - 1s 4ms/step - loss: 543.8150 - val_loss: 486.3482
Epoch 2/150
80/80 [==============================] - 0s 2ms/step - loss: 256.4229 - val_loss: 168.7667
Epoch 3/150
80/80 [==============================] - 0s 2ms/step - loss: 141.0736 - val_loss: 127.0370
Epoch 4/150
80/80 [==============================] - 0s 2ms/step - loss: 124.6489 - val_loss: 109.5187
Epoch 5/150
80/80 [==============================] - 0s 2ms/step - loss: 114.3285 - val_loss: 92.1651
Epoch 6/150
80/80 [==============================] - 0s 2ms/step - loss: 108.5375 - val_loss: 80.0352
Epoch 7/150
80/80 [==============================] - 0s 2ms/step - loss: 105.6313 - val_loss: 73.8658
Epoch 8/150
80/80 [==============================] - 0s 2ms/step - loss: 103.8833 - val_loss: 69.1125
Epoch 9/150
80/80 [==============================] - 0s 2ms/step - loss: 103.1798 - val_loss: 70.3567
Epoch 10/150
80/80 [==============================] - 0s 2ms/step - loss: 103.